Dispatcher tries to return used channels that are `<-chan interface{}` as `chan interface` and fails

# New Issue

## Context

## Expected Behavior

Tickers, orderbooks and balance data can be accessed through a subscription using a multiplexer using the `dispatch` package.
Subscribing to a UUID of a ticker/orderbook/account provides a 'Pipe' that holds a channel which is provided through sync.Pool. When unsubscribing from the item's updates through releasing the pipe, the used channel is returned to the pool. On a subsequent subscription, Pool.Get() doesn't return a New() item, but it pops a used channel and returns it to the caller.

## Current Behavior

Currently, when unsubscribing, a `<-chan interface{}` is put into the Pool, while subscribing tries to get a `chan interface{}`. As the pool holds used items as `any` types, a type assertion error is produced when accessing a used channel of type `<-chan interface{}` as a `chan interface{}` in dispatch/dispatch.go:245 

```
	ch, ok := getResult.(chan interface{})
	if !ok {
		return nil, errTypeAssertionFailure
	}
```


## Failure Information (for bugs)

### Steps to Reproduce

To reproduce, subscribe, unsubscribe and subscribe as per this (updated) test:

https://github.com/TaltaM/gocryptotrader/commit/387fca940b59801c2192eec4e6cc6b1479fd85f7

```
$ go test -v ./dispatch
=== RUN   TestGlobalDispatcher
--- PASS: TestGlobalDispatcher (0.00s)
<...snap...>
=== CONT  TestSubscribe
=== NAME  TestUnsubscribe
    dispatch_test.go:229: received: 'type assertion failure - <-chan interface {} is not a chan interface {}' but expected: '<nil>'
--- FAIL: TestUnsubscribe (0.00s)
<...snap...>
```

--- a/dispatch/dispatch.go
+++ b/dispatch/dispatch.go
@@ -241,8 +241,7 @@ func (d *Dispatcher) subscribe(id uuid.UUID) (chan interface{}, error) {
 	}
 
 	// Get an unused channel from the channel pool
-	getResult := d.outbound.Get()
-	ch, ok := getResult.(chan interface{})
+	ch, ok := d.outbound.Get().(chan interface{})
 	if !ok {
 		return nil, errTypeAssertionFailure
 	}

--- a/dispatch/dispatch.go
+++ b/dispatch/dispatch.go
@@ -218,7 +218,7 @@ func (d *Dispatcher) publish(id uuid.UUID, data interface{}) error {
 
 // Subscribe subscribes a system and returns a communication chan, this does not
 // ensure initial push.
-func (d *Dispatcher) subscribe(id uuid.UUID) (<-chan interface{}, error) {
+func (d *Dispatcher) subscribe(id uuid.UUID) (chan interface{}, error) {
 	if d == nil {
 		return nil, errDispatcherNotInitialized
 	}
@@ -241,7 +241,8 @@ func (d *Dispatcher) subscribe(id uuid.UUID) (<-chan interface{}, error) {
 	}
 
 	// Get an unused channel from the channel pool
-	ch, ok := d.outbound.Get().(chan interface{})
+	getResult := d.outbound.Get()
+	ch, ok := getResult.(chan interface{})
 	if !ok {
 		return nil, errTypeAssertionFailure
 	}
@@ -251,7 +252,7 @@ func (d *Dispatcher) subscribe(id uuid.UUID) (<-chan interface{}, error) {
 }
 
 // Unsubscribe unsubs a routine from the dispatcher
-func (d *Dispatcher) unsubscribe(id uuid.UUID, usedChan <-chan interface{}) error {
+func (d *Dispatcher) unsubscribe(id uuid.UUID, usedChan chan interface{}) error {
 	if d == nil {
 		return errDispatcherNotInitialized
 	}

--- a/dispatch/dispatch_test.go
+++ b/dispatch/dispatch_test.go
@@ -426,7 +426,7 @@ func TestMux(t *testing.T) {
 			return
 		}
 		errChan <- nil
-	}(pipe.C, errChan, &wg)
+	}(pipe.c, errChan, &wg)
 
 	wg.Wait()
 
@@ -506,7 +506,7 @@ func TestMuxPublish(t *testing.T) {
 		}
 	}(mux)
 
-	<-pipe.C
+	<-pipe.Channel()
 
 	// Shut down dispatch system
 	err = d.stop()

--- a/dispatch/dispatch_test.go
+++ b/dispatch/dispatch_test.go
@@ -183,7 +183,7 @@ func TestUnsubscribe(t *testing.T) {
 	}
 
 	// will return nil if not running
-	err = d.unsubscribe(nonEmptyUUID, make(<-chan interface{}))
+	err = d.unsubscribe(nonEmptyUUID, make(chan interface{}))
 	if !errors.Is(err, nil) {
 		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
 	}
@@ -193,7 +193,7 @@ func TestUnsubscribe(t *testing.T) {
 		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
 	}
 
-	err = d.unsubscribe(nonEmptyUUID, make(<-chan interface{}))
+	err = d.unsubscribe(nonEmptyUUID, make(chan interface{}))
 	if !errors.Is(err, errDispatcherUUIDNotFoundInRouteList) {
 		t.Fatalf("received: '%v' but expected: '%v'", err, errDispatcherUUIDNotFoundInRouteList)
 	}
@@ -203,7 +203,7 @@ func TestUnsubscribe(t *testing.T) {
 		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
 	}
 
-	err = d.unsubscribe(id, make(<-chan interface{}))
+	err = d.unsubscribe(id, make(chan interface{}))
 	if !errors.Is(err, errChannelNotFoundInUUIDRef) {
 		t.Fatalf("received: '%v' but expected: '%v'", err, errChannelNotFoundInUUIDRef)
 	}

--- a/dispatch/dispatch_test.go
+++ b/dispatch/dispatch_test.go
@@ -223,6 +223,16 @@ func TestUnsubscribe(t *testing.T) {
 	if !errors.Is(err, nil) {
 		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
 	}
+
+	ch2, err := d.subscribe(id)
+	if !errors.Is(err, nil) {
+		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
+	}
+
+	err = d.unsubscribe(id, ch2)
+	if !errors.Is(err, nil) {
+		t.Fatalf("received: '%v' but expected: '%v'", err, nil)
+	}
 }
 
 func TestPublish(t *testing.T) {

--- a/dispatch/dispatch_types.go
+++ b/dispatch/dispatch_types.go
@@ -72,7 +72,7 @@ type Mux struct {
 // Pipe defines an outbound object to the desired routine
 type Pipe struct {
 	// Channel to get all our lovely information
-	C chan interface{}
+	c chan interface{}
 	// ID to tracked system
 	id uuid.UUID
 	// Reference to multiplexer

--- a/dispatch/dispatch_types.go
+++ b/dispatch/dispatch_types.go
@@ -72,7 +72,7 @@ type Mux struct {
 // Pipe defines an outbound object to the desired routine
 type Pipe struct {
 	// Channel to get all our lovely information
-	C <-chan interface{}
+	C chan interface{}
 	// ID to tracked system
 	id uuid.UUID
 	// Reference to multiplexer

--- a/dispatch/mux.go
+++ b/dispatch/mux.go
@@ -42,7 +42,7 @@ func (m *Mux) Subscribe(id uuid.UUID) (Pipe, error) {
 }
 
 // Unsubscribe returns channel to the pool for the full signature set
-func (m *Mux) Unsubscribe(id uuid.UUID, ch <-chan interface{}) error {
+func (m *Mux) Unsubscribe(id uuid.UUID, ch chan interface{}) error {
 	if m == nil {
 		return errMuxIsNil
 	}

--- a/dispatch/mux.go
+++ b/dispatch/mux.go
@@ -38,7 +38,7 @@ func (m *Mux) Subscribe(id uuid.UUID) (Pipe, error) {
 		return Pipe{}, err
 	}
 
-	return Pipe{C: ch, id: id, m: m}, nil
+	return Pipe{c: ch, id: id, m: m}, nil
 }
 
 // Unsubscribe returns channel to the pool for the full signature set
@@ -83,5 +83,10 @@ func (m *Mux) GetID() (uuid.UUID, error) {
 
 // Release returns the channel to the communications pool to be reused
 func (p *Pipe) Release() error {
-	return p.m.Unsubscribe(p.id, p.C)
+	return p.m.Unsubscribe(p.id, p.c)
+}
+
+// Channel returns the Pipe's channel
+func (p *Pipe) Channel() <-chan interface{} {
+	return p.c
 }

--- a/engine/rpcserver.go
+++ b/engine/rpcserver.go
@@ -714,7 +714,7 @@ func (s *RPCServer) GetAccountInfoStream(r *gctrpc.GetAccountInfoRequest, stream
 	}()
 
 	for {
-		data, ok := <-pipe.C
+		data, ok := <-pipe.Channel()
 		if !ok {
 			return errDispatchSystem
 		}
@@ -2188,7 +2188,7 @@ func (s *RPCServer) GetExchangeOrderbookStream(r *gctrpc.GetExchangeOrderbookStr
 	}()
 
 	for {
-		data, ok := <-pipe.C
+		data, ok := <-pipe.Channel()
 		if !ok {
 			return errDispatchSystem
 		}
@@ -2273,7 +2273,7 @@ func (s *RPCServer) GetTickerStream(r *gctrpc.GetTickerStreamRequest, stream gct
 	}()
 
 	for {
-		data, ok := <-pipe.C
+		data, ok := <-pipe.Channel()
 		if !ok {
 			return errDispatchSystem
 		}
@@ -2326,7 +2326,7 @@ func (s *RPCServer) GetExchangeTickerStream(r *gctrpc.GetExchangeTickerStreamReq
 	}()
 
 	for {
-		data, ok := <-pipe.C
+		data, ok := <-pipe.Channel()
 		if !ok {
 			return errDispatchSystem
 		}

--- a/exchanges/account/account_test.go
+++ b/exchanges/account/account_test.go
@@ -213,7 +213,7 @@ func TestGetHoldings(t *testing.T) {
 		for i := 0; i < 2; i++ {
 			c := time.NewTimer(time.Second)
 			select {
-			case <-p.C:
+			case <-p.Channel():
 			case <-c.C:
 			}
 		}

